x3d-serializer: specular highlights and shape names

andreasplesch opened the pull request:
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

Rather subtle improvement to rendering by adding some specular highlighting similar to the default renderer. See #1234.

andreasplesch commented:
[Without highlights (current)](https://create3000.github.io/x_ite/playground/?url=https://gist.githubusercontent.com/andreasplesch/dc3595c5cf5198bedd7dbdb119a7a059/raw/ffb746984bcd31340ac9d2b2a88650c7ff37ab57/jscad.x3d)
[With highlights, this PR](https://create3000.github.io/x_ite/playground/?url=https://gist.githubusercontent.com/andreasplesch/dc3595c5cf5198bedd7dbdb119a7a059/raw/ffb746984bcd31340ac9d2b2a88650c7ff37ab57/basicColorsSpecular.x3d)

z3dev commented:
@andreasplesch this seems reasonable but i think colors and other material attributes should be provided at the shape level.

several of the serializers already support other attributes on shapes, for example 'name', 'id', etc.

if the other X3D material attributes are important then how about adding support for those?

https://doc.x3dom.org/developer/x3dom/nodeTypes/Material.html

andreasplesch commented:
Yes, the Material fields correspond to the shape level. Only diffuseColor is important for 3d printing, I believe. The other fields are important for screen rendering. The initial idea was to get closer to the rendering of the regl renderer by default which also uses specular highlights. Further refinement could be done in postprocessing outside of jscad. For that it would be useful to support a 'name' attribute which would correspond to the X3D DEF attribute.

It would be possible to look for object properties with the Material field names such as object.ambientIntensity and then use those for the x3d serialization. This way other serializers or the regl renderer could also use these object properties.

Should these properties be grouped in a material (or x3d) subobject: object.material.ambientIntensity ?
Should these properties be managed or propagated by jscad in operations which create new objects ? This seems out of scope at this point.

There are some decisions which then would follow:

- Use x3d defaults values, except for specularColor and shininess.
- geom3 uses color as diffuseColor. If both are given use diffuseColor.
- line geom uses color as emissiveColor. If both are given use emissiveColor.
- If line geom has properties other than color, include all in the material node. This is legal although x3d will only use emissive.
- Allow options for all fields to be used if object property does not exist.

So, it may suffice to support only a 'name' attribute and expect postprocessing outside of jscad. Adding support for all material fields as object attributes is also possible, potentially allowing use of those elsewhere although they are somewhat x3d specific.

I am fine with both options.

andreasplesch commented:
I added support for a .name attribute. There is no default option. Because X3D requires unique DEF names, I added a function which adds suffixes for Shapes which have identical .names. This could be left out and become a responsibility for the designer.

With regards to Material attributes, this may not be a priority but please let me know and I can add support.

z3dev commented on packages/io/x3d-serializer/src/index.js:
```diff
@@ -165,6 +168,24 @@ const convertGeom2 = (object, options) => {
+
+const shapeAttributes = (object, attributes = {}) => {
+  if (object.name) {
+    Object.assign(attributes, { DEF: uniqueDefName(object.name) })
```
The 'name' should be decided by the user. So, let's just use whatever the user provides.

andreasplesch replied:
ok, I can remove the unique function. Would it be possible to provide a warning, perhaps to the console, in case the same name is used for multiple objects ?

z3dev replied:
that sounds reasonable. by definition, are DEF (names) expected to be unique?

z3dev commented on packages/io/x3d-serializer/src/index.js:
```diff
@@ -180,14 +201,20 @@ const convertAppearance = (object, colorField, options) => {
+  if (colorField === 'diffuseColor') {
+    Object.assign(
+      materialFields,
+      { specularColor: '0.2 0.2 0.2', shininess: options.shininess })
```
Can something similar be done here for material attributes?

andreasplesch replied:
Ok, we can look for all Material attributes as outlined above.

z3dev replied:
Cool.

By the way, what ever becomes supported in the serializer should be supported in the deserializer. We should support the round trip...

andreasplesch replied:
I noticed that the deserializer uses obect.id for DEF names. Let me then use .id here as well
 rather than .name

andreasplesch replied:
Would material attributes in a sub property be appropriate: object.material.ambientIntensity ?

z3dev replied:
That seems reasonable. Please add some notes to the README file.

andreasplesch commented:
Yes, DEF names in x3d are required to be unique. x3d processors may
complain.

---on the phone---

On Mon, May 1, 2023, 2:38 AM Z3 Development ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In packages/io/x3d-serializer/src/index.js
> <https://github.com/jscad/OpenJSCAD.org/pull/1239#discussion_r1181419529>:
>
> > @@ -165,6 +168,24 @@ const convertGeom2 = (object, options) => {
>    return group
>  }
>
> +/*
> + * generate attributes for Shape node
> + */
> +
> +const shapeAttributes = (object, attributes = {}) => {
> +  if (object.name) {
> +    Object.assign(attributes, { DEF: uniqueDefName(object.name) })
>
> that sounds reasonable. by definition, are DEF (names) expected to be
> unique?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jscad/OpenJSCAD.org/pull/1239#discussion_r1181419529>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPCT266NBVHKWQJJDPVNULXD5K6TANCNFSM6AAAAAAXP6LHNU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

z3dev reviewed (commented):
@andreasplesch these changes look fine.

just two things...
- please run 'npm run lint' from the main directory and see if any issues are highlighted
- please set the status of your branch to NOT draft

andreasplesch commented:
Thanks for taking a look. I fixed the lint reports and requested review to convert from draft to PR. If  this is ready to be merged, I can address remaining items later in another PR.

z3dev reviewed (approved):
@andreasplesch YES! Thanks much for the cleanup.

All looks good!
Thanks for the changes. Looking forward to the next set of changes.
